### ISSUE TITLE
Memory leak in Concourse v5.4

### ISSUE THREAD

**Issue by kcmannem** [concourse/concourse#4150]
We don't know what exactly is happening. But we've observed this in our k8s helm deployment, after we upgraded to v5.4 we noticed this trend in web memory.

![Screen Shot 2019-07-23 at 10 56 43 AM](https://user-images.githubusercontent.com/9054726/61723502-43e06a80-ad3a-11e9-8624-005940799fdb.png)

**Comment by kcmannem** [comment]
I suspect its the parallel input streaming pr, where returned slices from calling a function are escaping and not being gced for some reason. We rolled our k8s deployment with the reverted commit and will provide an update here in a few hours to see if the leak is still occuring.

**Comment by kcmannem** [comment]
Here's how the mem has grown since the redeploy without parallel streams commit. It still looks like a very slow leak. I'm going to check back and update the picture in the morning. 
![Screen Shot 2019-07-23 at 10 23 15 PM](https://user-images.githubusercontent.com/9054726/61760125-78830f00-ad98-11e9-80c7-e85e7039cf9d.png)

We can see in the following graph that gc ticks have been occurring.
![Screen Shot 2019-07-23 at 10 25 08 PM](https://user-images.githubusercontent.com/9054726/61760221-db74a600-ad98-11e9-82df-2c345d88c9eb.png)

**Comment by kcmannem** [comment]
I also want to rule out the zstd change as then we'd also be seeing this behavior on the workers.

**Comment by gerhard** [comment]
We are seeing the same thing on a BOSH-deployed Concourse v5.4. Memory went from 18% to 21% in 20 hours, so the web memory leak is ~6MB/hour in our case.

![image](https://user-images.githubusercontent.com/3342/61781933-1d373800-adfd-11e9-984e-d4f9238ed519.png)

Maybe unrelated, but we are seeing a terrible leak in containers:

![image](https://user-images.githubusercontent.com/3342/61782014-3d66f700-adfd-11e9-9fb6-0f2b6e3c3a5f.png)

[This is a snapshot of all metrics, a day before & a day after we upgraded from v5.3 to v5.4](https://metrics.ci.rabbitmq.com/dashboard/snapshot/SDXyYZcUo1ljXB9GVAr8moUcRVhh2BYk?orgId=1)

![image](https://user-images.githubusercontent.com/3342/61782390-de55b200-adfd-11e9-95e1-dce508f655c9.png)

**Comment by kcmannem** [comment]
@gerhard do you have alot of resources that require creds in their configs?

**Comment by kcmannem** [comment]
I'd like to rule out any leaks introduced by go mod updates. I've done a diff between 5.3s go.mod and 5.4s. Only differences are zstd was introduced and baggage claim was bumped to 1.6

```
15a16
> 	github.com/DataDog/zstd v1.4.0
48c49
< 	github.com/concourse/baggageclaim v1.4.0
---
> 	github.com/concourse/baggageclaim v1.6.0
```

**Comment by kcmannem** [comment]
To really confirm the zstd leak, I'd like to do a memory inspection from the host level. We've done flame graphs using perf before and It'd be worthwhile to do the same for memory. http://www.brendangregg.com/FlameGraphs/memoryflamegraphs.html 

Hopefully its able to annotate different parts of the flame so we know exactly what's causing what.

**Comment by trucken** [comment]
I've observed the memory leakage in our deployment as well. We have roughly 80 pipelines and each got 16 resources, of which 11 require credentials. We got 2 workers handling all the resources, the magenta and blue in Worker Volumes diagram. And as you can see we have low activity during the night and the virtual memory and resident memory still increases, albeit slower than during active periods.

![Concourse_53_54](https://user-images.githubusercontent.com/3066579/61866608-c6516180-aed5-11e9-8ebf-4d26ce17c310.png)

Hope it can help you in the right direction.

**Comment by gerhard** [comment]
> @gerhard do you have alot of resources that require creds in their configs? Or are most of the resources from public places like public github repos.

Yes, we do have a lot of resources that require credentials. Because we have many large pipelines, it's hard to give you an exact number. Is there a quick way of finding this out? grepping through yaml files would be too slow and inefficient.

> To really confirm whats leaking (zstd or something else), I'd like to do a memory inspection from the host level.

That sounds most reasonable. In case I don't have time to do this today, I will look into it tomorrow.

**Comment by kcmannem** [comment]
@trucken @gerhard Appreciate the help. We ran a test overnight of a docker composed concourse, with some lightweight pipelines. We saw memory grow from 1% -> 4%. We're running the same test again today but put zstd streaming into a loop(streaming 4000 times instead of just once) to exhaust that function and see if its actually leaking. We are also going to isolate that function in a smaller test and see if we can reproduce it. I'm leaning towards it not having to do with creds interpolation, a part of the code we refactored during the same time as zstd's merge.

**Comment by bhutkovskyysos** [comment]
I am been monitoring this issue and just wanted to throw my experience here. 

We use a monorepo and I have a relatively large pipeline (about 30 jobs, one per app), which get build and pushed to AWS ECR. I noticed that with load when all services are building in parallel in docker resources I see auth errors "no basic auth credentials". 

I also have an older version of concourse 3.14 and never experienced this before. It is probably related to the leakage mentioned above.

**Comment by kcmannem** [comment]
// NewReader returns a new io.ReadCloser that will decompress data from the
// underlying reader.  If a dictionary is provided to NewReaderDict, it must
// not be modified until Close is called.  It is the caller's responsibility
// to call Close, which frees up C objects.

**Comment by samgurtman** [comment]
Is this going to result in a new release?

**Comment by kcmannem** [comment]
@samgurtman sometime next week is the plan. The patches have to go through topgun which can take a little while.

**Comment by gerhard** [comment]
@kcmannem great work, thank you for seeing this through so promptly!

**Comment by puellanivis** [comment]
Thanks for locating the bug and and explanation of what was causing it, it’s a great learning point.

But as a bit of an aside, while I was trying to track down a memory leak for some colleagues, I noticed a codepath where the code was not closing a `gzip.Reader`, and noted to them that [according to the `compress/gzip` library](https://golang.org/pkg/compress/gzip/#NewReader) there was the same caveat that the caller needs to call `Close`.

However, after testing, and as noted above, if one does not call `Close` (at least for the currently supported versions of Go) `gzip.Reader`s are still GCed. So, I could not get it to actually leak memory, which meant that for my colleagues, I had to move on looking for a cause of their memory leak.

But I’ll definitely put this issue in my pocket, and call for good `Close` sanitation with any and all compression library `Reader`, even if it does not actually leak memory. It’s hard to know when an implementations might change from under you, and suddenly your compression library’s `Reader` is no longer GC’ed automatically…

**Comment by gaigaslab-operations** [comment]
Is there going to be a patch release for this fix?  If so, what's the ETA on that release?

**Comment by kcmannem** [comment]
@gaigaslab-operations this week. Everything's finished for the release.

**Comment by pivotal-jamie-klassen** [comment]
Fixed in #4165, #4185, #4191 on master, duplicated as #4182 and #4199 on release/5.4.x.

**Comment by gerhard** [comment]
I'm intrigued: will our 2 web nodes exhaust memory before the next v5.4.x patch ships?

<img width="1601" alt="Screenshot 2019-08-07 at 13 53 16" src="https://user-images.githubusercontent.com/3342/62624324-bcd6e900-b91a-11e9-80e5-52caadd3a3df.png">
